[1.11.2] Added support for Village capabilities

Madsthunder opened the pull request:
This PR adds the fields and methods required for Capabilities into the Village class, similar to Entity, TileEntity, ItemStack, and World.

Madsthunder committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.11.x' into village-capabilities

Madsthunder committed:
Added capabilities to villages

Madsthunder committed:
Fixed imports

Madsthunder committed:
Slightly reduced the patch size

Madsthunder committed:
Remove extra line at the top of ForgeEventFactory

Madsthunder committed:
Fixed type in ForgeEventFactory

Madsthunder committed:
Changed method reference in Village

shadowfacts commented:
Why is this necessary/useful?

mezz commented:
I also think this seems very strange. What are the village capabilities used for?

Madsthunder commented:
@shadowfacts This is useful because it allows a modder to store extra data about a village. For example, if a modder needs a unique identifier for a village they can do so by creating a capability that serializes/deserializes a UUID. This would remove the need for modders to try to determine which village a villager is from by assuming from NBT, because the village would have a unique identifier.

Madsthunder commented:
@mezz The village capabilities could be used for multiple things, like if a modder wanted to add economy into the game, the village could represent a nation, the modder could then add a capability called Wealth or something along those lines  and assign it to the appropriate Village.

mezz commented on patches/minecraft/net/minecraft/village/Village.java.patch:
```diff
@@ -0,0 +1,72 @@
+             }
+         }
++        if (this.capabilities != null && p_82690_1_.func_74764_b("ForgeCaps")) this.capabilities.deserializeNBT(p_82690_1_.func_74775_l("ForgeCaps"));
++
```
please remove extra newline here

mezz commented on patches/minecraft/net/minecraft/village/Village.java.patch:
```diff
@@ -0,0 +1,72 @@
+ 
+     public Village()
+     {
++        this(null);
```
This seems slightly questionable? May be better to just set the capabilities here too. I'm open to other opinions though.

mezz commented on patches/minecraft/net/minecraft/village/Village.java.patch:
```diff
@@ -0,0 +1,72 @@
+             this.field_75590_b = p_i1674_3_;
+         }
+     }
++
```
Could use a 
`/* ======================================== FORGE START =====================================*/`
section for good measure

Madsthunder committed:
Apply suggestions from mezz (1)

Madsthunder commented:
Changed the constructor call to Village(World) in Village() to a gatherCapabilites call, removed the extra line at the end of writeVillageToNBT and added a Forge Start/End surrounding the new methods.

mezz commented:
Can you make a small forge debug mod that uses this capability, to ensure it is working as intended and nothing else is needed?
There are several in `net.minecraftforge.debug`, if you need any help getting set up there is a [tutorial video here](https://www.youtube.com/watch?v=pLWQk6ed56Q).

Madsthunder commented:
@mezz Would you mind if I integrated it into the existing TestCapabilityMod?

mezz commented:
Sure, as long as it's not interacting with the existing stuff there it should be fine.

LatvianModder commented:
Forge should figure out a way to store unique IDs for - Worlds/dimensions, villages and the server...

Madsthunder committed:
Fixed Village AttachCapabilitesEvent not firing correctly

Madsthunder committed:
Added Villages to the TestCapabilityMod

Madsthunder commented:
@mezz Fixed the AttachCapabilitiesEvent not firing correctly and added Villages to the TestCapabilityMod. The mod also checks every tick for players entering and exiting villages and prints it to the chat.

**EDIT:**
One thing to note is that this only works on server for some reason

rguy96 commented:
@mezz Any idea when the next batch of PRs are being pulled?

mezz commented:
@rguy96 Probably after PAX South, so in February.

LexManos commented:
between moving, and PAX, I'm tied up till the 3rd. Also this one needs to be looked into quite a but. I'm not a particular fan of just sticking caps on every possible thing. I do think this needs a few more concrete explained if its use case.

mezz commented:
@Madsthunder can you address Lex's comment by giving a few more concrete examples of uses for this event?

Madsthunder commented:
Other than the examples above, I can't really think of any off the top of my head, and the use cases I have for it are a bit too complex to put into one comment...

mezz commented:
@Madsthunder can you update this to resolve conflicts?

I think this is a good way to attach extra data to villages, which is not simple to do otherwise.
I'm not exactly sure capabilities are the perfect fit because I can't easily imagine how mods would share the capabilities. However it's not a very invasive patch and the implementation is simple.
I think it's up to Lex to look into this a bit more when he has more time to explain.

Madsthunder committed:
Merge branch '1.11.x' into village-capabilities

mezz commented on src/test/java/net/minecraftforge/test/TestCapabilityMod.java:
```diff
@@ -186,7 +190,51 @@ public void attachTileEntity(AttachCapabilitiesEvent<TileEntity> event)
+                if(village.hasCapability(TEST_CAP, null))
+                {
+                    boolean playerInRadius = false;
+                    for(EntityPlayer player : players)
```
this `for` loop needs braces

mezz commented on patches/minecraft/net/minecraft/village/Village.java.patch:
```diff
@@ -135,3 +170,36 @@
++    private net.minecraftforge.common.capabilities.CapabilityDispatcher capabilities;
++    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
++    {
++        if (getCapability(capability, facing) != null)
```
`hasCapability` is meant to be a faster cheaper check than `getCapability` so it should not call `getCapability`.

LexManos commented on patches/minecraft/net/minecraft/village/Village.java.patch:
```diff
@@ -24,19 +45,20 @@
 +    }
 +
-+    public int getPlayerReputation(UUID p_82684_1_)
++    public int getPlayerReputation(UUID playerName)
```
Dunno how you messed this up, but fix it.

Madsthunder committed:
Suggestions (5.0)

Madsthunder committed:
Suggestions (5.1)
